core/internal/sentry/sentry.go

```diff
@@ -0,0 +1,176 @@
+package sentry
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"log/slog"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/getsentry/sentry-go"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/wandb/wandb/core/internal/version"
+	"github.com/wandb/wandb/core/pkg/observability"
+)
+
+type Params struct {
+	DSN     string
+	Commit  string
+	LRUSize int
+}
+
+type Client struct {
+	DSN    string
+	Commit string
+	mu     sync.Mutex
+	Recent *lru.Cache
+}
+
+var recentErrorDuration = time.Minute * 5
+
+// RemoveBottomFrames modifies the stack trace by checking the file name of the bottom-most 3 frames
+// and removing them if they are internal to core
+func RemoveBottomFrames(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
+	for i, exception := range event.Exception {
+		if exception.Stacktrace == nil {
+			continue
+		}
+		frames := exception.Stacktrace.Frames
+		framesLen := len(frames)
+		// for the recovered panics, the bottom-most 3 frames of the stacktrace
+		// will come from sentry.go and logging.go, so we remove them
+		if framesLen < 3 {
+			continue
+		}
+		for j := framesLen - 1; j >= framesLen-3; j-- {
+			frame := frames[j]
+			// TODO: think of a better way to do this without hard-coding the file names
+			//  this is a hack to remove the bottom-most 3 frames that are internal to core
+			if strings.HasSuffix(frame.AbsPath, "sentry.go") || strings.HasSuffix(frame.AbsPath, "logging.go") {
+				frames = frames[:j]
+			} else {
+				break
+			}
+		}
+		event.Exception[i].Stacktrace.Frames = frames
+	}
+	return event
+}
+
+// New initializes the sentry client.
+func New(params Params) *Client {
+	if params.LRUSize == 0 {
+		params.LRUSize = 100
+	}
+	cache, err := lru.New(params.LRUSize)
+	if err != nil {
+		slog.Error("failed to create LRU cache", "err", err)
+		return nil
+	}
+
+	// If the DSN is not set, the client is effectively disabled.
+	s := &Client{
+		DSN:    params.DSN,
+		Commit: params.Commit,
+		Recent: cache,
+	}
+
+	err = sentry.Init(sentry.ClientOptions{
+		Dsn:              s.DSN,
+		AttachStacktrace: true,
+		Release:          version.Version,
+		Dist:             s.Commit,
+		BeforeSend:       RemoveBottomFrames,
+	})
+
+	if err != nil {
+		slog.Error("sentry.Init failed", "err", err)
+	}
+
+	if s.DSN != "" {
+		slog.Debug("sentry.Init succeeded", "dsn", s.DSN)
+	} else {
+		slog.Debug("sentry is disabled")
+	}
+
+	return s
+}
+
+func (s *Client) shouldCapture(err error) bool {
+	s.mu.Lock()
```

Comment:
The lock is mostly (but not totally) redundant now... but i can see why you would want it, if you have a flood of the same events from different parallel contexts.   Just wanted to make sure this was intentional to keep this here.  I dont know if this is a case we care about, so up to you if you want to keep the lock since the lru is safe already.

Reply:
good point, it's redundant now, so rm'ed